File: test_scrolls.dart
import 'dart:io';
import 'package:supabase_flutter/supabase_flutter.dart';

void main() async {
  print('Loading Env...');
  final envVars = await _loadEnv('.env');
  final url = envVars['SUPABASE_URL']!;
  final anonKey = envVars['SUPABASE_ANON_KEY']!;

  print('Connecting to Supabase...');
  await Supabase.initialize(url: url, anonKey: anonKey);
  final db = Supabase.instance.client;

  print('Testing Profiles for free users...');
  try {
    final limitRes = await db
        .from('profiles')
        .select('id, email, scrolls_count, daily_free_scrolls')
        .limit(3);
    print(limitRes);

    print('\nCalling use_scroll RPC directly for first user...');
    final rpcRes =
        await db.rpc('use_scroll', params: {'user_id': limitRes[0]['id']});
    print('RPC Result: $rpcRes');

    final afterRes = await db
        .from('profiles')
        .select('scrolls_count, daily_free_scrolls')
        .eq('id', limitRes[0]['id']);
    print('After RPC: $afterRes');
  } catch (e) {
    print('Error: $e');
  }

  exit(0);
}

Future<Map<String, String>> _loadEnv(String path) async {
  final file = File(path);
  final lines = await file.readAsLines();
  final varMap = <String, String>{};
  for (var line in lines) {
    if (line.trim().isEmpty || line.startsWith('#')) continue;
    final parts = line.split('=');
    if (parts.length >= 2) {
      final key = parts[0].trim();
      final value = parts.sublist(1).join('=').trim();
      varMap[key] = value;
    }
  }
  return varMap;
}

File: test_scrolls.js
const { createClient } = require("@supabase/supabase-js");
const fs = require("fs");
require("dotenv").config();

const supabaseUrl = process.env.SUPABASE_URL;
const supabaseKey = process.env.SUPABASE_SERVICE_ROLE_KEY;
const supabase = createClient(supabaseUrl, supabaseKey);

async function testScrolls() {
  console.log("Fetching test profile...");
  const { data, error } = await supabase
    .from("profiles")
    .select(
      "id, email, scrolls_count, daily_free_scrolls, last_scroll_refreshed_date",
    )
    .limit(1);

  if (error) {
    console.error(error);
    return;
  }

  const user = data[0];
  console.log("Current State:", user);

  console.log("Attempting to use_scroll via RPC...");
  const { data: rpcData, error: rpcError } = await supabase.rpc("use_scroll", {
    user_id: user.id,
  });

  if (rpcError) {
    console.error("RPC Error:", rpcError);
  } else {
    console.log("RPC Success. Scroll deducted?", rpcData);
  }

  const { data: afterData } = await supabase
    .from("profiles")
    .select("scrolls_count, daily_free_scrolls")
    .eq("id", user.id)
    .single();

  console.log("State After RPC:", afterData);
}

testScrolls();
